### PR TITLE
More precise pool shares

### PR DESCRIPTION
**Pull request by gbalabasquer**

### CODE REVIEW

File: src/D3MHub.sol
Comment (gbalabasquer) [inline src/D3MHub.sol:290]: Instead of `WAD` I think we could just do 1 in the precision diff between DAI and the token. For Aave 1 wei should be enough, for Compound I think the margin should be 10^10. But doing that different per `ilk` would require ro add that number somehow in the `pool` or `plan` and I don't think it is probably worth it. 1 WAD is not much difference to really care.

File: src/pools/D3MAavePool.sol
Comment (talbaneth) [inline src/pools/D3MAavePool.sol:185]: Logic looks good.

If we care much about gas here we can:
* Consider keeping `end` in storage (fileable).
* Caching `exited` in the beginning and doing something like:  `exited = _exited +  wad`

File: src/pools/ID3MPool.sol
Comment (talbaneth) [inline src/pools/ID3MPool.sol:54]: Maybe we should update the documentation a bit.
I guess at least the @notice part should be something like "Exit proportional amount of shares".

File: src/tests/stubs/D3MTestPool.sol
Comment (talbaneth) [inline src/tests/stubs/D3MTestPool.sol:152]: Can you explain this change please.

File: src/tests/pools/D3MCompoundPool.t.sol
Comment (talbaneth) [inline src/tests/pools/D3MCompoundPool.t.sol:220]: We might want a test that does 2 exits with an elapsing blocks (time in aave) period between them.
We want to make sure all dai and cdai were handed out.

File: src/pools/D3MAavePool.sol
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol]: - I started with a fileable `end` but then when I realized it could come from the `hub` change to the actual one. I know is more gas intensive but at the same time will save us complication for initial set up and in case of `End` migration.
Anyway not that strong feeling if you prefer the other way.
- Done

File: src/tests/stubs/D3MTestPool.sol
Comment (gbalabasquer) [reply src/tests/stubs/D3MTestPool.sol:152]: I had a test failing due this. The correct behaviour of `maxWithdraw` should be the minimum between the shares of DAI owned by our pool contract in the third party pool and the total available DAI in that moment in the whole pool. So basically now is doing what the real pools do.

Comment (gbalabasquer) [reply src/tests/stubs/D3MTestPool.sol:152]: This was the test failing: `test_wind_limited_by_pool_loss` and it makes sense as share decreased from our balance.
Which btw I'm now renaming to: `test_unwind_due_to_by_pool_loss`

File: src/pools/D3MAavePool.sol
Comment (talbaneth) [reply src/pools/D3MAavePool.sol]: Ok, I'm also fine with the way `end` is fetched from the hub.
Resolving.

File: src/tests/stubs/D3MTestPool.sol
Comment (talbaneth) [reply src/tests/stubs/D3MTestPool.sol:152]: perfect

File: src/pools/D3MAavePool.sol
Comment (kmbarry1) [inline src/pools/D3MAavePool.sol:187]: There's an implicit invariant here, right? That `end().Art(ilk)` needs to be at least as much as the exitable `ink`. Couldn't that be violated in some repayment scenarios?

Comment (kmbarry1) [reply src/pools/D3MAavePool.sol:187]: Not suggesting we need to block on it here, just calibrating my understanding for unit testing.

File: src/tests/pools/D3MCompoundPool.t.sol
Comment (gbalabasquer) [reply src/tests/pools/D3MCompoundPool.t.sol:220]: Added new tests.

File: src/D3MHub.sol
Comment (gbalabasquer) [reply src/D3MHub.sol:290]: Other thing that realized and discussed with @talbaneth, adding up `WAD` means that a malicious value of `currentAssets > maxinuint256 - WAD` will make a potential unwinding to stop.
We agreed that caging or making the plan inactive (which won't require gov delay) it is probably more than enough protection for it (as the conditional will jump before than evaluating it). cc @kmbarry1

**Review by talbaneth — APPROVED** [review]
LGTM

File: src/D3MHub.sol
Comment (kmbarry1) [reply src/D3MHub.sol:290]: Good edge case to call out but I doubt it would be hit in practice, and if someone is able to manipulate the `currentAssets` value (or if it has a bug) action would be necessary anyway. Don't think the side effects here are terribly severe, either.

Comment (gbalabasquer) [reply src/D3MHub.sol:290]: Yeah seems we are all align regarding this. Thanks for your feedback!

File: src/tests/integration/D3MAave.t.sol
Comment (iamchrissmith) [inline src/tests/integration/D3MAave.t.sol:983]:
```suggestion
        // User can exit and get the aDAI
```

File: src/pools/D3MAavePool.sol
Comment (iamchrissmith) [inline src/pools/D3MAavePool.sol:97]: rather than passing this in for each pool in the constructor (kind of hard to enforce in our current interface pattern), how would you feel about adding it to the `exit` args?

File: src/tests/integration/D3MCompound.t.sol
Comment (iamchrissmith) [inline src/tests/integration/D3MCompound.t.sol:1026]: why switch to `1e18` from `WAD`?

File: src/tests/pools/D3MAavePool.t.sol
Comment (iamchrissmith) [inline src/tests/pools/D3MAavePool.t.sol:258]: Am I understanding this correctly? `vm.prank` do not change the value of `address(this)` it only changes `msg.sender`?

File: src/tests/pools/D3MAavePool.t.sol
Comment (iamchrissmith) [inline src/tests/pools/D3MAavePool.t.sol:250]: since we now don't expect this to be called except during ES, maybe we can remove this test? I think it is a hold over from when `transfer` was for upgrading too

File: src/tests/pools/D3MCompoundPool.t.sol
Comment (iamchrissmith) [inline src/tests/pools/D3MCompoundPool.t.sol:220]: same comment as on Aave, I think we can remove this test and just leave the `caged` one

File: Makefile
Comment (iamchrissmith) [inline Makefile:4]: will this pass along arguments if passed?

File: src/tests/pools/D3MAavePool.t.sol
Comment (gbalabasquer) [reply src/tests/pools/D3MAavePool.t.sol:258]: Right, that's how I understand it.

File: src/tests/integration/D3MCompound.t.sol
Comment (gbalabasquer) [reply src/tests/integration/D3MCompound.t.sol:1026]: That's how it was already in the cDai tests

File: src/tests/pools/D3MAavePool.t.sol
Comment (gbalabasquer) [reply src/tests/pools/D3MAavePool.t.sol:250]: To be honest I think the one that doesn't make sense is the vat caged one. This is a unit test for the `pool` and the `vat` status has nothing to do with this method.

File: Makefile
Comment (gbalabasquer) [reply Makefile:4]: I just checked and it does, eventually we should directly remove it though.

File: src/pools/D3MAavePool.sol
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol:97]: I considered this option but eventually felt it was more an implementation detail for these two pools, that's why ended up passing the `ilk` in the constructor. If we consider this is actually something to generalize, then yeah we could move to the function param.

File: src/tests/pools/D3MAavePool.t.sol
Comment (iamchrissmith) [reply src/tests/pools/D3MAavePool.t.sol:250]: actually, you're right.  Lets remove the cage one?

File: src/tests/integration/D3MCompound.t.sol
Comment (iamchrissmith) [reply src/tests/integration/D3MCompound.t.sol:1026]: it looked like it was just a few places in those, so i created this to change to WAD for consistency: https://github.com/makerdao/dss-direct-deposit/pull/88

File: src/tests/pools/D3MCompoundPool.t.sol
Comment (gbalabasquer) [reply src/tests/pools/D3MCompoundPool.t.sol:220]: Removed the `cage` on as discussed in the other comment.

File: src/tests/integration/D3MCompound.t.sol
Comment (gbalabasquer) [reply src/tests/integration/D3MCompound.t.sol]: Merged!

**Review by iamchrissmith — APPROVED** [review]
looks good to me. before you merge, we should check with @godsflaw and @naszam to see if they would prefer to review these changes here or in the main branch the red team is reviewing

**Comment by godsflaw** [comment]
You guys going to rug me and merge this?  I will have to start over anyway (if I'm following the strict format of the matrix), so I'm okay with that.  It already happened to me once.  I would rather review this change in place than wonder if it invalidates my review.

**Comment by kmbarry1** [comment]
> You guys going to rug me and merge this? I will have to start over anyway (if I'm following the strict format of the matrix), so I'm okay with that. It already happened to me once. I would rather review this change in place than wonder if it invalidates my review.

In any case I've been waiting for this PR to get finalized before writing exhaustive tests...happy to write them against this branch if we want to wait on the merge.

**Pull request merged**
